### ISSUE TITLE
Support SslMasterKeyHandler in NettyServerBuilder

### ISSUE THREAD

**Issue by nucz** [grpc/grpc-java#7199]
I have a problem with adding SslMasterKeyHandler#newWireSharkSslMasterKeyHandler on the server side. I tried to do this with channelFactory:
```
builder.channelFactory(() -> {
          final ServerChannel ch = GrpcUtil.getDefaultServerChannelFactory().newChannel();
          final ChannelPipeline pipeline = ch.pipeline();
          pipeline.addLast(SslMasterKeyHandler.newWireSharkSslMasterKeyHandler());
          return ch;
        })
```
and later with ProtocolNegotiator:
```
final SslContext sslContext = GrpcSslContexts.configure(sslClientContextBuilder).build();
final io.grpc.netty.ProtocolNegotiator negotiator = ProtocolNegotiators.serverTls(sslContext);
builder.protocolNegotiator(tlsNegotiator(negotiator);
```

In the first case, the setting had no effect. In the second case, the server worked incorrectly (it stopped accepting connections, even without a handler, just with a negotiator wrapper). 

What should be the proper implementation?

**Comment by voidzcy** [comment]
I don't think this is the correct usage of WireSharkSslMasterKeyHandler. You would only need to enable it by setting the system property `-Dio.netty.ssl.masterKeyHandler=true`, no modification for the pipeline is needed. See details in its [Javadoc](https://netty.io/4.1/api/io/netty/handler/ssl/SslMasterKeyHandler.html#SYSTEM_PROP_KEY).

**Comment by ejona86** [comment]
@voidzcy, I don't think that system property would work as-is. That only works if gRPC included SslMasterKeyHandler in the pipeline already.

Changing the `channelFactory()` didn't work because that's the listening channel factory. It produces children for each connection and those are what you are wanting to modify. The `protocolNegotiator()` snippet looks pretty mangled, such that it is unclear what it would look like once you actually got it to compile.

I don't think there is a proper implementation currently. You could maybe hack something with the channelFactory(), but it would be a big hack. It isn't really safe for you to change the pipeline without gRPC being aware; we've found Netty pipelines to be very brittle.

**Comment by ejona86** [comment]
@nucz, would it be fair to call this an enhancement request asking for SslMasterKeyHandler support?

It seems we could check the system property `io.netty.ssl.masterKeyHandler` just like Netty and if it is true we add the handler to the pipeline. I'd prefer not having it on the pipeline all the time, to avoid any potential performance/behavior confusion. I'm disappointed it doesn't remove itself from the pipeline after the handshake completes; I wonder if it doesn't to allow supporting rekeying (I don't remember what is supported there).

**Comment by nucz** [comment]
@ejona86 Yes, that would be great, to have SslMasterKeyHandler support. I found the solution using negotiators, but it is far from ideal and is version sensitive.

**Comment by Nnadozie** [comment]
@ejona86 I would like to take a stab at this, but off the top of your head, please what's your estimate for this unit of work?

**Comment by ejona86** [comment]
@Nnadozie, should be like 3 lines of code. You'd basically check the system property, and if set, create and add the additional handler [within ServerTlsHandler](https://github.com/grpc/grpc-java/blob/d52b359631ee1ac46f47d72d1015e1dd6ce9312c/netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java#L267).

**Comment by Nnadozie** [comment]
Thanks @ejona86! Working on a PR now.

**Comment by huangqiangxiong** [comment]
@ejona86 @Nnadozie I suggest that this feature can also be used on the client side, because sometimes you can only debug on the client side (because the server is provided by others).

In addition, I recommend to extend .sslContext(SslContext sslContext) to .sslContext(SslContext sslContext, boolean allowExportMasterKey). To do pipeline.addLast(SslMasterKeyHandler.newWireSharkSslMasterKeyHandler()), only while both the allowExportMasterKe is true and the system property "io.netty.ssl.masterKeyHandler" is "true". Because sometimes there are several grpc clients working in one program, we may only track one client's traffic.

**Comment by huangqiangxiong** [comment]
Hi, @Nnadozie
Are you still working on this feature? If you don't have time, please tell me, I may make a try.

**Comment by Nnadozie** [comment]
@huangqiangxiong I haven't had time lately so please go ahead.

**Comment by huangqiangxiong** [comment]
@ejona86 
I modified the code according to your early suggestion (checking the system property, and if set, create and add the additional handler within ServerTlsHandler) because I find the ClientTlsHandler/ServerTlsHandler are not only created through NettyChannelBuilder or NettyServerBuilder. So I give up my original idea. Only change ClientTlsHandler/ServerTlsHandler is more simple and enough for debug.

**Comment by ejona86** [comment]
Reopening since #7724 was reverted by #7792.

**Comment by huangqiangxiong** [comment]
I find an alternative way to export TLS key materials in Java that is using jsslkeylog and replacing boringssl sslProvider with JDK building security Provider.

I think some people may get help from it, so record it here https://gitlab.com/wireshark/wireshark/-/wikis/How-to-Export-TLS-Master-keys-of-gRPC#using-jsslkeylog

**Comment by vinodhabib** [comment]
Started Analysis and going through the previous comments

**Comment by kannanjgithub** [comment]
The [issue](https://github.com/netty/netty/issues/10957) filed with Netty for master key export not working with TLS 1.3 is still open and huangqiangxiong has given us a way of exporting the key material by using JDK built-in security provider. There is no action from our side until the Netty issue is fixed for TLS 1.3.
